server/gui/main.py

```diff
@@ -0,0 +1,212 @@
+        self.thread_pool = QThreadPool()
+        self.cef_widget = None
+        self.data_widget = None
+        self.server = cellxgeneServer(self)
+        self.server.setup_app()
+        self.runServer()
+        self.setWindowTitle("cellxgene")
+
+        # Strong focus - accepts focus by tab & click
+        self.setFocusPolicy(Qt.StrongFocus)
+        self.setupLayout()
+        self.setupMenu()
+
+    def setupLayout(self):
+        self.resize(WIDTH, HEIGHT)
+        self.cef_widget = CefWidget(self)
+        self.data_widget = LoadWidget(self)
+        self.stacked_layout = QStackedLayout()
+        self.stacked_layout.addWidget(self.data_widget)
+        self.stacked_layout.addWidget(self.cef_widget)
+        main_layout = QVBoxLayout()
+        main_layout.setContentsMargins(0, 0, 0, 0)
+        main_layout.setSpacing(0)
+        main_layout.addLayout(self.stacked_layout)
+        frame = QFrame()
+        frame.setLayout(main_layout)
+        self.setCentralWidget(frame)
+
+        if WINDOWS:
+            # On Windows with PyQt5 main window must be shown first
+            # before CEF browser is embedded, otherwise window is
+            # not resized and application hangs during resize.
+            self.show()
+
+        # Browser can be embedded only after layout was set up
+        self.cef_widget.embedBrowser()
+
+        if LINUX:
+            # On Linux with PyQt5 the QX11EmbedContainer widget is
+            # no more available. An equivalent in Qt5 is to create
+            # a hidden window, embed CEF browser in it and then
+            # create a container for that hidden window and replace
+            # cef widget in the layout with the container.
+            self.container = QWidget.createWindowContainer(
+                self.cef_widget.hidden_window, parent=self)
+            stacked_layout.addWidget(self.container, 1, 0)
+
+    def setupMenu(self):
+        main_menu = self.menuBar()
+        file_menu = main_menu.addMenu('File')
+        load_action = QAction("Load file...", self)
+        load_action.setStatusTip("Load file")
+        load_action.setShortcut("Ctrl+O")
+        load_action.triggered.connect(self.showLoad)
+        file_menu.addAction(load_action)
+
+    def showLoad(self):
+        self.stacked_layout.setCurrentIndex(0)
+
+    def closeEvent(self, event):
+        # Close browser (force=True) and free CEF reference
+        if self.cef_widget.browser:
+            self.cef_widget.browser.CloseBrowser(True)
+            self.clearBrowserReferences()
+
+    def runServer(self):
+        worker = ServerRunWorker(self.server.app, host=self.server.host, port=self.server.port)
+        self.thread_pool.start(worker)
+
+    def clearBrowserReferences(self):
+        # Clear browser references that you keep anywhere in your
+        # code. All references must be cleared for CEF to shutdown cleanly.
+        self.cef_widget.browser = None
+
+
+# TODO make central location for methods?
```

Comment:
thumbs up on the generalization - don't think these kinds of general config belong in the Qt app.  We will want them exposed to other wrappers as well (notebook, Qt, CLI, ...).

Reply:
should I do this at another point? or get it in this PR?

Reply:
OK with me if you do it progressively (ie, not in this PR)

Reply:
just did it, it was easy